### PR TITLE
test_e2e can now run parallel e2e test on  multiple clusters with the same -i and -b configuration

### PR DESCRIPTION
**Pull request by qabpea**
Moved all binaries to bin directory.
Changed installation to check if {kind, kubectl, ginko] is in the directory instead of checking if they are in the existing PATH.

Added -s (suffix) argument, the suffix will be appended to the e2e directory name and also to the kind cluster names.
Added -n (test_run_count) argument, this will control how many clusters that will be created in parallel and how many tests that will execute in parallel for the same -i and -b combination

### CODE REVIEW

**Comment by dougsland** [comment]
Hello @qabpea , nice patch but I am facing the error below:

` ./hack/test_e2e.sh -i ipv4 -b iptables -n 10 -d`
```
ERROR: failed to create cluster: docker run error: command "docker run --hostname kpng-e2e-ipv4-iptables-worker --name kpng-e2e-ipv4-iptables-worker --label io.x-k8s.kind.role=worker --privileged --security-opt seccomp=unconfined --security-opt apparmor=unconfined --tmpfs /tmp --tmpfs /run --volume /var --volume /lib/modules:/lib/modules:ro --detach --tty --label io.x-k8s.kind.cluster=kpng-e2e-ipv4-iptables --net kind --restart=on-failure:1 --init=false docker.io/kindest/node:v1.22.2" failed with error: exit status 125
Command Output: docker: Error response from daemon: Conflict. The container name "/kpng-e2e-ipv4-iptables-worker" is already in use by container "877f07341f703899c1b2847a0ab6abc2eafb0ec7d2a9d98c40ff8cc95e20506b". You have to remove (or rename) that container to be able to reuse that name.
See 'docker run --help'.
 ✗ Preparing nodes 📦 📦 📦 
ERROR: failed to create cluster: docker run error: command "docker run --hostname kpng-e2e-ipv4-iptables-worker2 --name kpng-e2e-ipv4-iptables-worker2 --label io.x-k8s.kind.role=worker --privileged --security-opt seccomp=unconfined --security-opt apparmor=unconfined --tmpfs /tmp --tmpfs /run --volume /var --volume /lib/modules:/lib/modules:ro --detach --tty --label io.x-k8s.kind.cluster=kpng-e2e-ipv4-iptables --net kind --restart=on-failure:1 --init=false docker.io/kindest/node:v1.22.2" failed with error: exit status 125
Command Output: docker: Error response from daemon: Conflict. The container name "/kpng-e2e-ipv4-iptables-worker2" is already in use by container "63cc1b6772dbe5d012a51d184dc73df000df6a19327386cdab7cb57fd0a78846". You have to remove (or rename) that container to be able to reuse that name.

```

**Comment by qabpea** [comment]
Ok my bad, I will fix that, for now, try to add a - p and set a suffix.
________________________________

**Comment by qabpea** [comment]
@dougsland bug fixed plus a little more functionality added

**Review by nehaLohia27 — APPROVED** [review]

**Comment by jayunit100** [comment]
thanks doug for testing -> @qabpea can you add 

```
   run: ./hack/test_e2e.sh -i ${{ env.IP_FAMILY }} -b ${{ env.BACKEND }} -c -n 1
```

to the .github/actions , then CI will exersize this, and we can merge :)

**Comment by dougsland** [comment]
@qabpea  Looks better but still how some errors. Is it something local only (in my env)? 

```
# ./hack/test_e2e.sh -i ipv4 -b iptables -n 10 -d
+==================================================================+
		Starting KPNG E2E testing
+==================================================================+
[ PASSED ] Detected Container Engine: docker
[ PASSED ] Image build and tag kpng:test is set.

Downloading kind ...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100    98  100    98    0     0    573      0 --:--:-- --:--:-- --:--:--   573
100   655  100   655    0     0   2243      0 --:--:-- --:--:-- --:--:--  2243
100 6660k  100 6660k    0     0  6960k      0 --:--:-- --:--:-- --:--:-- 13.9M
[ PASSED ] The kind tool is set.

Downloading kubectl ...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   154  100   154    0     0   1092      0 --:--:-- --:--:-- --:--:--  1092
100 44.7M  100 44.7M    0     0  12.3M      0  0:00:03  0:00:03 --:--:-- 14.0M
[ PASSED ] The kubectl tool is set.

Downloading ginkgo and e2e.test ...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   154  100   154    0     0   1509      0 --:--:-- --:--:-- --:--:--  1509
100  270M  100  270M    0     0  13.4M      0  0:00:20  0:00:20 --:--:-- 13.8M
[ PASSED ] The tools ginko and e2e.test have been set up.

+=====================================================================================+
		Running parallel KPNG E2E tests in background on 10 kind clusters.
+=====================================================================================+

+==================================================================+
		Test Report from running test on 10 clusters.
+==================================================================+
Summary report from cluster "1" in directory: "/root/test/kpng/temp/e2e-1"
Summary report from cluster "2" in directory: "/root/test/kpng/temp/e2e-2"
Summary report from cluster "3" in directory: "/root/test/kpng/temp/e2e-3"
Summary report from cluster "4" in directory: "/root/test/kpng/temp/e2e-4"
Summary report from cluster "5" in directory: "/root/test/kpng/temp/e2e-5"
Summary report from cluster "6" in directory: "/root/test/kpng/temp/e2e-6"
Summary report from cluster "7" in directory: "/root/test/kpng/temp/e2e-7"
Summary report from cluster "8" in directory: "/root/test/kpng/temp/e2e-8"
Summary report from cluster "9" in directory: "/root/test/kpng/temp/e2e-9"
Summary report from cluster "10" in directory: "/root/test/kpng/temp/e2e-10"

Occurence	Failure
```

`kpng/temp/e2e-1/output.log `
```
 1 
  2 Preparing to setup kpng-e2e-ipv4-iptables-1 cluster ...
  3 Creating cluster "kpng-e2e-ipv4-iptables-1" ...
  4  • Ensuring node image (docker.io/kindest/node:v1.22.2) 🖼   ...
  5  ✓ Ensuring node image (docker.io/kindest/node:v1.22.2) 🖼
  6  • Preparing nodes 📦 📦 📦   ...
  7  ✓ Preparing nodes 📦 📦 📦
  8  • Writing configuration 📜  ...
  9  ✓ Writing configuration 📜
 10  • Starting control-plane 🕹️   ...
 11  ✗ Starting control-plane 🕹️
 12 ERROR: failed to create cluster: failed to init node with kubeadm: command "docker exec --privileged kpng-e2e-ipv4-iptables-1-control-plane kubeadm init --skip-phases=prefli
 13 
 14 Command Output: Error response from daemon: Container 0d10b8fdae13b3fd6b87d9d4635996af30d91ca27eb8b0efce0a5c85da252581 is not running
 15 [ ^[[31mFAILED^[[0m ] cannot create kind cluster kpng-e2e-ipv4-iptables-1                                                                              
```
`kpng/temp/e2e-2/output.log`
```
6 I0130 13:18:14.718396     213 round_trippers.go:454] GET https://kpng-e2e-ipv4-iptables-2-control-plane:6443/healthz?timeout=10s  in 0 milliseconds
307 I0130 13:18:15.218542     213 round_trippers.go:454] GET https://kpng-e2e-ipv4-iptables-2-control-plane:6443/healthz?timeout=10s  in 0 milliseconds
308 I0130 13:18:15.718810     213 round_trippers.go:454] GET https://kpng-e2e-ipv4-iptables-2-control-plane:6443/healthz?timeout=10s  in 0 milliseconds
309 I0130 13:18:16.218891     213 round_trippers.go:454] GET https://kpng-e2e-ipv4-iptables-2-control-plane:6443/healthz?timeout=10s  in 0 milliseconds
310 I0130 13:18:16.719139     213 round_trippers.go:454] GET https://kpng-e2e-ipv4-iptables-2-control-plane:6443/healthz?timeout=10s  in 0 milliseconds
311 I0130 13:18:17.218506     213 round_trippers.go:454] GET https://kpng-e2e-ipv4-iptables-2-control-plane:6443/healthz?timeout=10s  in 0 milliseconds
312 I0130 13:18:17.718676     213 round_trippers.go:454] GET https://kpng-e2e-ipv4-iptables-2-control-plane:6443/healthz?timeout=10s  in 0 milliseconds
313 I0130 13:18:18.218995     213 round_trippers.go:454] GET https://kpng-e2e-ipv4-iptables-2-control-plane:6443/healthz?timeout=10s  in 0 milliseconds
314 I0130 13:18:18.718391     213 round_trippers.go:454] GET https://kpng-e2e-ipv4-iptables-2-control-plane:6443/healthz?timeout=10s  in 0 milliseconds
315 I0130 13:18:19.218138     213 round_trippers.go:454] GET https://kpng-e2e-ipv4-iptables-2-control-plane:6443/healthz?timeout=10s  in 0 milliseconds
316 I0130 13:18:19.718360     213 round_trippers.go:454] GET https://kpng-e2e-ipv4-iptables-2-control-plane:6443/healthz?timeout=10s  in 0 milliseconds
317 I0130 13:18:20.218560     213 round_trippers.go:454] GET https://kpng-e2e-ipv4-iptables-2-control-plane:6443/healthz?timeout=10s  in 0 milliseconds
318 I0130 13:18:20.718995     213 round_trippers.go:454] GET https://kpng-e2e-ipv4-iptables-2-control-plane:6443/healthz?timeout=10s  in 0 milliseconds
319 I0130 13:18:21.218190     213 round_trippers.go:454] GET https://kpng-e2e-ipv4-iptables-2-control-plane:6443/healthz?timeout=10s  in 0 milliseconds
320 I0130 13:18:21.718490     213 round_trippers.go:454] GET https://kpng-e2e-ipv4-iptables-2-control-plane:6443/healthz?timeout=10s  in 0 milliseconds
321 [kubelet-check] It seems like the kubelet isn't running or healthy.
322 [kubelet-check] The HTTP call equal to 'curl -sSL http://localhost:10248/healthz' failed with error: Get "http://localhost:10248/healthz": dial tcp [::1]:10248: connect: connection refused.
323 
324         Unfortunately, an error has occurred:
325                 timed out waiting for the condition
326 
327         This error is likely caused by:
328                 - The kubelet is not running
329                 - The kubelet is unhealthy due to a misconfiguration of the node in some way (required cgroups disabled)
330 
331         If you are on a systemd-powered system, you can try to troubleshoot the error with the following commands:
332                 - 'systemctl status kubelet'
333                 - 'journalctl -xeu kubelet'
334 
335         Additionally, a control plane component may have crashed or exited when started by the container runtime.
336         To troubleshoot, list all containers using your preferred container runtimes CLI.
337 
338         Here is one example how you may list all Kubernetes containers running in cri-o/containerd using crictl:
339                 - 'crictl --runtime-endpoint unix:///run/containerd/containerd.sock ps -a | grep kube | grep -v pause'
340                 Once you have found the failing container, you can inspect its logs with:
341                 - 'crictl --runtime-endpoint unix:///run/containerd/containerd.sock logs CONTAINERID'
342 
343 couldn't initialize a Kubernetes cluster
344 k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/init.runWaitControlPlanePhase
345         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/init/waitcontrolplane.go:116
346 k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow.(*Runner).Run.func1
347         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow/runner.go:234
348 k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow.(*Runner).visitAll
349         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow/runner.go:421
350 k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow.(*Runner).Run
351         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow/runner.go:207
352 k8s.io/kubernetes/cmd/kubeadm/app/cmd.newCmdInit.func1
353         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/cmd/init.go:153
354 k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).execute
355         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:852
356 k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).ExecuteC
357         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:960
358 k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).Execute
359         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:897
360 k8s.io/kubernetes/cmd/kubeadm/app.Run
361         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/kubeadm.go:50
362 main.main
363         _output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/kubeadm.go:25
364 runtime.main
365         /usr/local/go/src/runtime/proc.go:225
366 runtime.goexit
367         /usr/local/go/src/runtime/asm_amd64.s:1371
368 error execution phase wait-control-plane
369 k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow.(*Runner).Run.func1
370         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow/runner.go:235
371 k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow.(*Runner).visitAll
372         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow/runner.go:421
373 k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow.(*Runner).Run
374         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow/runner.go:207
375 k8s.io/kubernetes/cmd/kubeadm/app/cmd.newCmdInit.func1
376         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/cmd/init.go:153
377 k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).execute
378         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:852
379 k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).ExecuteC
380         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:960
381 k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).Execute
382         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:897
383 k8s.io/kubernetes/cmd/kubeadm/app.Run
384         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/kubeadm.go:50
385 main.main
386         _output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/kubeadm.go:25
387 runtime.main
388         /usr/local/go/src/runtime/proc.go:225
389 runtime.goexit
390         /usr/local/go/src/runtime/asm_amd64.s:1371
391 [ ^[[31mFAILED^[[0m ] cannot create kind cluster kpng-e2e-ipv4-iptables-2
```


`kpng/temp/e2e-3/output.log`
```
0130 13:18:23.032887     228 round_trippers.go:454] GET https://kpng-e2e-ipv4-iptables-3-control-plane:6443/healthz?timeout=10s  in 0 milliseconds
314 I0130 13:18:23.533048     228 round_trippers.go:454] GET https://kpng-e2e-ipv4-iptables-3-control-plane:6443/healthz?timeout=10s  in 0 milliseconds
315 I0130 13:18:24.033274     228 round_trippers.go:454] GET https://kpng-e2e-ipv4-iptables-3-control-plane:6443/healthz?timeout=10s  in 0 milliseconds
316 I0130 13:18:24.533638     228 round_trippers.go:454] GET https://kpng-e2e-ipv4-iptables-3-control-plane:6443/healthz?timeout=10s  in 0 milliseconds
317 I0130 13:18:25.034068     228 round_trippers.go:454] GET https://kpng-e2e-ipv4-iptables-3-control-plane:6443/healthz?timeout=10s  in 0 milliseconds
318 I0130 13:18:25.533375     228 round_trippers.go:454] GET https://kpng-e2e-ipv4-iptables-3-control-plane:6443/healthz?timeout=10s  in 0 milliseconds
319 I0130 13:18:26.033653     228 round_trippers.go:454] GET https://kpng-e2e-ipv4-iptables-3-control-plane:6443/healthz?timeout=10s  in 0 milliseconds
320 I0130 13:18:26.533819     228 round_trippers.go:454] GET https://kpng-e2e-ipv4-iptables-3-control-plane:6443/healthz?timeout=10s  in 0 milliseconds
321 [kubelet-check] It seems like the kubelet isn't running or healthy.
322 [kubelet-check] The HTTP call equal to 'curl -sSL http://localhost:10248/healthz' failed with error: Get "http://localhost:10248/healthz": dial tcp [::1]:10248: connect: connection refused.
323 
324         Unfortunately, an error has occurred:
325                 timed out waiting for the condition
326 
327         This error is likely caused by:
328                 - The kubelet is not running
329                 - The kubelet is unhealthy due to a misconfiguration of the node in some way (required cgroups disabled)
330 
331         If you are on a systemd-powered system, you can try to troubleshoot the error with the following commands:
332                 - 'systemctl status kubelet'
333                 - 'journalctl -xeu kubelet'
334 
335         Additionally, a control plane component may have crashed or exited when started by the container runtime.
336         To troubleshoot, list all containers using your preferred container runtimes CLI.
337 
338         Here is one example how you may list all Kubernetes containers running in cri-o/containerd using crictl:
339                 - 'crictl --runtime-endpoint unix:///run/containerd/containerd.sock ps -a | grep kube | grep -v pause'
340                 Once you have found the failing container, you can inspect its logs with:
341                 - 'crictl --runtime-endpoint unix:///run/containerd/containerd.sock logs CONTAINERID'
342 
343 couldn't initialize a Kubernetes cluster
344 k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/init.runWaitControlPlanePhase
345         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/init/waitcontrolplane.go:116
346 k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow.(*Runner).Run.func1
347         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow/runner.go:234
348 k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow.(*Runner).visitAll
349         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow/runner.go:421
350 k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow.(*Runner).Run
351         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow/runner.go:207
352 k8s.io/kubernetes/cmd/kubeadm/app/cmd.newCmdInit.func1
353         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/cmd/init.go:153
354 k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).execute
355         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:852
356 k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).ExecuteC
357         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:960
358 k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).Execute
359         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:897
360 k8s.io/kubernetes/cmd/kubeadm/app.Run
361         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/kubeadm.go:50
362 main.main
363         _output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/kubeadm.go:25
364 runtime.main
365         /usr/local/go/src/runtime/proc.go:225
366 runtime.goexit
367         /usr/local/go/src/runtime/asm_amd64.s:1371
368 error execution phase wait-control-plane
369 k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow.(*Runner).Run.func1
370         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow/runner.go:235
371 k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow.(*Runner).visitAll
372         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow/runner.go:421
373 k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow.(*Runner).Run
374         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow/runner.go:207
375 k8s.io/kubernetes/cmd/kubeadm/app/cmd.newCmdInit.func1
376         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/cmd/init.go:153
377 k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).execute
378         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:852
379 k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).ExecuteC
380         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:960
381 k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).Execute
382         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:897
383 k8s.io/kubernetes/cmd/kubeadm/app.Run
384         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/kubeadm.go:50
385 main.main
386         _output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/kubeadm.go:25
387 runtime.main
388         /usr/local/go/src/runtime/proc.go:225
389 runtime.goexit
390         /usr/local/go/src/runtime/asm_amd64.s:1371
391 [ ^[[31mFAILED^[[0m ] cannot create kind cluster kpng-e2e-ipv4-iptables-3
```

**Comment by qabpea** [comment]
The - d stops the program after the kind clusters are created. I need an if statement checkin before calling print report
________________________________

**Comment by qabpea** [comment]
Yes, the -d flag (devel_flag) is used to skip running the tests.
I missed to skip the printout in the parallel case, my bad.
It is fixed now and there is a printout that makes this clear

**Comment by dougsland** [comment]
> Yes, the -d flag (devel_flag) is used to skip running the tests. I missed to skip the printout in the parallel case, my bad. It is
> fixed now and there is a printout that makes this clear

That's a quick way to generate a bunch of test env, great stuff @qabpea 
```
# kind get clusters
kpng-e2e-ipv4-iptables
kpng-e2e-ipv4-iptables-1
kpng-e2e-ipv4-iptables-10
kpng-e2e-ipv4-iptables-2
kpng-e2e-ipv4-iptables-3
kpng-e2e-ipv4-iptables-4
kpng-e2e-ipv4-iptables-5
kpng-e2e-ipv4-iptables-6
kpng-e2e-ipv4-iptables-7
kpng-e2e-ipv4-iptables-8
kpng-e2e-ipv4-iptables-9
```

**Comment by dougsland** [comment]
It seems much clear output @qabpea 
```
root@douglas-PowerEdge-R610:~/kpng# ./hack/test_e2e.sh -i ipv4 -b iptables -n 3
+==================================================================+
		Starting KPNG E2E testing
+==================================================================+
[ PASSED ] Detected Container Engine: docker
[ PASSED ] Image build and tag kpng:test is set.
[ PASSED ] The kind tool is set.
[ PASSED ] The kubectl tool is set.
[ PASSED ] The tools ginko and e2e.test have been set up.

+=====================================================================================+
		Running parallel KPNG E2E tests in background on 3 kind clusters.
+=====================================================================================+
+==================================================================+
		Test Report from running test on 3 clusters.
+==================================================================+
Summary report from cluster "1" in directory: "/root/kpng/temp/e2e-1"
Ran 285 of 6432 Specs in 720.856 seconds
SUCCESS! -- 285 Passed | 0 Failed | 0 Pending | 6147 Skipped
Summary report from cluster "2" in directory: "/root/kpng/temp/e2e-2"
Ran 285 of 6432 Specs in 756.824 seconds
FAIL! -- 284 Passed | 1 Failed | 0 Pending | 6147 Skipped
Summary report from cluster "3" in directory: "/root/kpng/temp/e2e-3"
Ran 285 of 6432 Specs in 668.105 seconds
FAIL! -- 284 Passed | 1 Failed | 0 Pending | 6147 Skipped

Occurence	Failure
      2 [Fail] [sig-network] Conntrack [It] should be able to preserve UDP traffic when server pod cycles for a ClusterIP service 
		 /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/onsi/ginkgo/internal/leafnodes/runner.go:113
```

File: hack/test_e2e.sh [hunk 1]
```diff
@@ -641,9 +793,8 @@ if ! [[ "${backend}" =~ ^(iptables|nft|ipvs)$ ]]; then
 fi
 
 if [[ -n "${ip_family}" && -n "${backend}" ]]; then
-   main "${ip_family}" "${backend}" "${ci_mode}" "${e2e_dir}" "${bin_dir}" "${dockerfile}"
+    main "${ip_family}" "${backend}" "${ci_mode}" "${e2e_dir}" "${bin_dir}" "${dockerfile}" "${suffix}" "$test_run_count"
```
Comment (dougsland) [inline hack/test_e2e.sh]: additional space?

File: hack/test_e2e.sh [hunk 2]
```diff
@@ -492,31 +509,157 @@ function clean_artifacts {
     pass_message "make sure to safe your result before the next run."
 }
 
-function set_e2e_dir {
+
+function apply_ipvx_fixes {
     ###########################################################################
     # Description:                                                            #
-    # Set E2E directory                                                       #
+    # apply ipvx fixe                                                         #
```
Comment (dougsland) [inline hack/test_e2e.sh:554]: typo: fixes ?

**Comment by dougsland** [comment]
Small request: Can we add some words into `--help`  or other location regarding `not-kpng` ? How to use it? How to compare?
To my eyes: lgtm

File: hack/test_e2e.sh (hunk 1 above)
Comment (qabpea) [reply hack/test_e2e.sh]: aligning with code in other if statements

File: hack/test_e2e.sh (hunk 2 above)
Comment (qabpea) [reply hack/test_e2e.sh:554]: yes

**Comment by qabpea** [comment]
I need to add a description in a separate doc how all the new flags can be used together. I will add a note in the help printout  regarding not-kpng

**Comment by dougsland** [comment]
@qabpea let's add the remaining requests in different patch. This one is already XL :)

**Pull request merged**